### DISCUSSION TITLE
How to configure Go custom registry using "host rule detection from env" feature

### DISCUSSION THREAD

**Discussion by celian-garcia** [renovatebot/renovate#17658] [Request Help] [answered]
Hello, 
We have a Jenkins job running every hour the renovate docker image.
We have some private bitbucket repositories and would like to activate go.mod dependency updates.

Several solutions has been identified to do so, but we would like to know why we can't use the renovate one or if we are missing something.

## Solution 0 : Renovate one (Our preferred one but not working)
https://docs.renovatebot.com/golang/#custom-registry-support-and-authentication

This is our preferred one but we would like to do it using the host rules detection by env var and we struggle to do it

We tried the following options
### Option 0.1 and 0.2
```
RENOVATE_DETECT_HOST_RULES_FROM_ENV=true 
BITBUCKET_<MY_URL>_TOKEN 
```
-> The bitbucket is well in the white list to perform git config update but it does not generate the host rule
```
RENOVATE_DETECT_HOST_RULES_FROM_ENV=true 
<MY_URL>_TOKEN
```
-> It does not generate the host rule
> **edit**: Reasons of why options 0.1 and 0.2 are not taken has been answered by @viceice here :
> we only support this for datasources
> 
> https://github.com/renovatebot/renovate/blob/2df8967ae27851e79d7c665483456825a21a6e6d/lib/workers/global/config/parse/host-rules-from-env.ts#L12
> 
> 
> https://github.com/renovatebot/renovate/blob/2df8967ae27851e79d7c665483456825a21a6e6d/lib/workers/global/config/parse/host-rules-from-env.ts#L26
> 
> https://docs.renovatebot.com/modules/datasource/#supported-datasources

### Option 0.3
```
RENOVATE_DETECT_HOST_RULES_FROM_ENV=true 
GO_<MY_URL>_TOKEN 
```
-> It generates the rule but go host type is not in the whitelist to perform git config udpate 

## Solution 1 : Update GIT_CONFIG_ (working but not in all cases)
```sh
GIT_CONFIG_KEY_0=url.https://${BITBUCKET_USERNAME}:${BITBUCKET_PASSWORD}@${BITBUCKET_URL}
```

## Solution 2 : Update netrc (working)
```sh
echo "machine ${BITBUCKET_URL} login ${BITBUCKET_USERNAME} password ${BITBUCKET_PASSWORD}" > ${HOME}/.netrc
```

**Comment by rarkins** [comment]
For which datasource do you need this authentication for?

**Reply by celian-garcia** [reply]
This is for Golang dependency updates. So that it can be authenticated when doing go get ./... on private deps.
The datasource then is go ? git ? both ? ^^

**Reply by celian-garcia** [reply]
I'd say git

**Reply by rarkins** [reply]
For your solution 0 attempts, can you look at the debug logs when Renovate starts to confirm if/which hostRules are detected from env?

**Reply by celian-garcia** [reply]
Yep that's what I did to know if hostRules are detected from env. 

I can see the message only when I set GO_ host type. 
Messages like that 

``DEBUG: Converting GO_<MY_HOST>_TOKEN into a global host rule``

Otherwise, with GIT_, or BITBUCKET_ or BITBUCKET__SERVER_, it doesn't display the message.

**Reply by celian-garcia** [reply]
~~Wait trying to verify again error messages, I re-run the pipeline setting up 
``GO_<MY_HOST>_TOKEN=${BITBUCKET_USERNAME}:${BITBUCKET_PASSWORD}`` 
And it seems to work ?? the go.sum is well created~~
**edit**: nevermind it doesn't

**Reply by viceice** [reply]
we only support this for datasources

https://github.com/renovatebot/renovate/blob/2df8967ae27851e79d7c665483456825a21a6e6d/lib/workers/global/config/parse/host-rules-from-env.ts#L12
https://github.com/renovatebot/renovate/blob/2df8967ae27851e79d7c665483456825a21a6e6d/lib/workers/global/config/parse/host-rules-from-env.ts#L26

**Reply by celian-garcia** [reply]
Okay thanks @viceice, that would answer the discussion's original question then yes it is limited, to I guess this list right ?
https://docs.renovatebot.com/modules/datasource/#supported-datasources

**Reply by viceice** [reply]
yes

**Reply by celian-garcia** [reply]
~~But then I'm wondering if ``GO_<MY_HOST>_TOKEN=${BITBUCKET_USERNAME}:${BITBUCKET_PASSWORD}`` was supposed to trigger the necessary git config update or if I am hacking it in a way. As it is not really what the Renovate documentation is saying about go.mod custom registry 
https://docs.renovatebot.com/golang/#custom-registry-support-and-authentication~~
**edit**: nevermind it doesn't

**Reply by celian-garcia** [reply]
Thanks both of you for the fast replies. I realized that my question was not reflecting my need, so I updated the title, and the description a bit to integrate the answer of @viceice. 

I hope it's not too messy, but we still don't manage to configure a custom registry from env variable. 
Dont't you think it would solve the issue to just add "go" in this list?
https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/gomod/artifacts.ts#L57-L65

**Comment by celian-garcia** [comment, marked as answer]
Hello, I would like to answer my own question. I updated to latest version 37.107.0 and the solution 0 works well now 
 "GO_<MY_HOST>_TOKEN=${BITBUCKET_USERNAME}:${BITBUCKET_PASSWORD}"

Thank you, we can close this issue
